performances/irobot-benchmark/src/irobot-benchmark.cpp

```diff
@@ -26,6 +26,68 @@
 
 using namespace std::chrono_literals;
 
+template <typename NodeT>
+void init_ros_system(int argc, char** argv,
+                    const irobot_benchmark::Options &options,
+                    const std::string &topology_json,
+                    pid_t &pid,
+                    const std::string &resources_output_path,
+                    const std::string &events_output_path,
+                    const std::string &latency_all_output_path,
+                    const std::string &latency_total_output_path)
+{
+    // Start resources logger
+    performance_test::ResourceUsageLogger ru_logger(resources_output_path);
+
+    ru_logger.start(std::chrono::milliseconds(options.resources_sampling_per_ms));
+
+    rclcpp::init(argc, argv);
+
+    performance_test::System<NodeT> ros2_system(static_cast<performance_test::ExecutorType>(options.executor));
```

Comment:
I made some generic comments above, however I'm not fully convinced by this approach, even though I understand that the coexistence between regular nodes and lifecycle nodes is not trivial.

I think that using templates is not the best solution, as the vast majority of functions that are part of the System and Factory classes probably don't care about the type of node that you use.

I would prefer to have the NodeType passed as an argument as we do for the executor type.

```
performance_test::System ros2_system(static_cast<performance_test::ExecutorType>(options.executor));
auto factory = performance_test::TemplateFactory(options.ipc, options.ros_params);
auto nodes_vec = factory.parse_topology_from_json(topology_json, options.tracking_options, static_cast<performance_test::NodeType>(options.node));
ros2_system.add_node(nodes_vec);
```

@mauropasse can you please have a quick look? Which approach do you thinks is the best?

Reply:
Sorry, I haven't had time to check how lifecycle nodes differ from regular nodes in terms of how to use them in the perf framework. I had to add some minor templated functions for the executor type, if that could be replicated also with lifecycle nodes that'd be nice! But I ignore how much this approach could complicate the implementation.

Reply:
I will look into using function overload or perhaps templated functions (not the class), though my concern was initially holding different containers (e.g. vector of Node or LifecycleNode) and having to add logic everywhere to check which type.